haystack/generator/transformers.py

```diff
@@ -0,0 +1,223 @@
+        self.generator_type = generator_type
+        self.num_beams = num_beams
+        self.embed_title = embed_title
+        self.prefix = prefix
+        self.retriever = retriever
+
+        if use_gpu and torch.cuda.is_available():
+            self.device = torch.device("cuda")
+        else:
+            self.device = torch.device("cpu")
+
+        self.tokenizer = RagTokenizer.from_pretrained(model_name_or_path)
+
+        if self.generator_type == RAGeneratorType.SEQUENCE:
+            raise NotImplementedError("RagSequenceForGeneration is not implemented yet")
+            # TODO: Enable when transformers have it. Refer https://github.com/huggingface/transformers/issues/7905
+            # Also refer refer https://github.com/huggingface/transformers/issues/7829
+            # self.model = RagSequenceForGeneration.from_pretrained(model_name_or_path)
+        else:
+            self.model = RagTokenForGeneration.from_pretrained(model_name_or_path)
+
+    # Copied cat_input_and_doc method from transformers.RagRetriever
+    # Refer section 2.3 of https://arxiv.org/abs/2005.11401
+    def _cat_input_and_doc(self, doc_title: str, doc_text: str, input_string: str, prefix: Optional[str]):
+        if doc_title.startswith('"'):
+            doc_title = doc_title[1:]
+        if doc_title.endswith('"'):
+            doc_title = doc_title[:-1]
+        if prefix is None:
+            prefix = ""
+        out = (prefix + doc_title + self.model.config.title_sep + doc_text + self.model.config.doc_sep +
+               input_string).replace("  ", " ")
+
+        return out
+
+    # Copied postprocess_docs method from transformers.RagRetriever and modified
+    def _get_contextualized_inputs(self, texts: List[str], question: str, titles: Optional[List[str]] = None,
+                                   return_tensors: str = "pt"):
+
+        titles_list = titles if self.embed_title and titles is not None else [""] * len(texts)
+        prefix = self.prefix if self.prefix is not None else self.model.config.generator.prefix
+
+        rag_input_strings = [
+            self._cat_input_and_doc(
+                doc_title=titles_list[i],
+                doc_text=texts[i],
+                input_string=question,
+                prefix=prefix
+            )
+            for i in range(len(texts))
+        ]
+
+        contextualized_inputs = self.tokenizer.generator.batch_encode_plus(
+            rag_input_strings,
+            max_length=self.model.config.max_combined_length,
+            return_tensors=return_tensors,
+            padding="max_length",
+            truncation=True,
+        )
+
+        return contextualized_inputs["input_ids"].to(self.device), \
+               contextualized_inputs["attention_mask"].to(self.device)
+
+    def embed_passage_in_tensor(self, docs: List[Document], embeddings: List[Optional[numpy.ndarray]]) -> torch.Tensor:
+
+        # If of document missing embedding, then need embedding for all the documents
```

Comment:
typo

Reply:
Also please suggest proper name for `embed_passage_in_tensor`

Reply:
Sorry missed your comment below